client/src/main/java/io/pravega/client/stream/impl/TxnKeepAlive.java

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * TxnKeepAlive is used to send KeepAlives to renew the transaction lease for active transactions.
+ * It invokes io.pravega.client.stream.impl.Controller#pingTransaction() on the controller to renew the lease of active
+ * transactions. Incase of a controller instance not being reachable the controller client takes care of retrying on a
+ * different controller instance see io.pravega.client.stream.impl.ControllerResolverFactory for details.
+ */
+@Slf4j
+public class TxnKeepAlive implements AutoCloseable {
+    private static final double KEEP_ALIVE_INTERVAL_FACTOR = 0.7; //keepalive interval = factor * txn lease time.
+
+    private final Stream stream;
+    private final Controller controller;
+    private final long txnLeaseMillis;
+    private final long pingIntervalMillis;
+    private final ScheduledExecutorService executor;
+    private final List<UUID> txnList = new CopyOnWriteArrayList<>();
+    private final AtomicBoolean isStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> pingTxnTask;
+
+    TxnKeepAlive(EventWriterConfig config, Stream stream, Controller controller, ScheduledExecutorService
+            internalExecutor) {
+        this.txnLeaseMillis = config.getTransactionTimeoutTime();
+        this.pingIntervalMillis = getKeepAliveInterval(txnLeaseMillis);
+        this.stream = stream;
+        this.controller = controller;
+        this.executor = internalExecutor;
+    }
+
+    void startTxnKeepAlive(UUID txnID) {
+        txnList.add(txnID);
+        startPeriodicPingTxn();
+    }
+
+    void stopTxnKeepAlive(UUID txnID) {
+        txnList.remove(txnID);
+    }
+
+    private long getKeepAliveInterval(long txnLeaseMillis) {
+        double keepAliveInterval = txnLeaseMillis * KEEP_ALIVE_INTERVAL_FACTOR;
+        if (keepAliveInterval < TimeUnit.SECONDS.toMillis(10)) {
+            log.warn("Transaction Keep Alive interval is less than 10 seconds(lower bound)");
+        }
+        //Ping interval cannot be less than KeepAlive task interval of 10seconds.
+        return Math.max(TimeUnit.SECONDS.toMillis(10), (long) keepAliveInterval);
+    }
+
+    @Synchronized
+    private void startPeriodicPingTxn() {
+        if (!isStarted.get()) {
+            log.info("Starting Transaction keep alive at an interval of {}ms ", this.pingIntervalMillis);
+            pingTxnTask = executor.scheduleAtFixedRate(this::pingTransactions, 10, this.pingIntervalMillis,
+                    TimeUnit.MILLISECONDS);
+            isStarted.set(true);
+        }
+    }
+
+    /*
+        Ping all the transactions present in the list. Controller client performs retries in case of a failures. On a
+         failure the particular transaction it is removed from the ping list.
+     */
+    private void pingTransactions() {
+        Map<UUID, CompletableFuture<Void>> pingResult =
+                txnList.stream().collect(Collectors.toMap(Function.identity(),
```

Comment:
since this is all done in future without a wait, there could be multiple pingTransactions that could have submitted ping futures for the same transaction concurrently. 
This seems slightly wasteful.
Instead of collecting all transactions in a single keepAlive, why not have schedules posted for them individually. And it doesnt even have to be a schedule.
It could be something like:
Futures.loop(() -> stopPingingForTx(T1), () -> FutureHelpers.delayedFuture(pingTxn(T1), delay))
This will ensure that you dont have to bundle all txns together. .there will also be exactly one pending ping request per txn.
and you can add error handling on pingTxn to make sure retries are done even if pings fail (because of some intermittent network issue).

Reply:
`java.util.concurrent.ScheduledExecutorService#scheduleAtFixedRate` ensures that executions will not concurrently execute.